app/models/concerns/profile_policy_scoring.rb

```diff
@@ -9,4 +9,34 @@ def score(host: nil)
 
     ((scores = results.pluck(:score)).sum / scores.size).to_f
   end
+
+  def update_counters!
+    # rubocop:disable Rails/SkipsModelValidations
+    update_columns(
+      total_host_count: hosts.count,
+      test_result_host_count: latest_supported_policy_test_result_hosts.count,
+      compliant_host_count: latest_supported_policy_test_result_hosts
+        .count { |host| compliant?(host) },
+      unsupported_host_count: latest_unsupported_policy_test_result_hosts.count
+    )
+    # rubocop:enable Rails/SkipsModelValidations
+  end
+
+  private
+
+  def latest_supported_policy_test_result_hosts
+    ::Host.where(id: latest_policy_test_results.supported.select(:host_id))
+  end
+
+  def latest_unsupported_policy_test_result_hosts
+    ::Host.where(id: latest_policy_test_results
+      .supported(false).select(:host_id))
+  end
+
+  def latest_policy_test_results
+    ::TestResult
+      .where(id: test_results.pluck(:id))
+      .or(::TestResult.where(profile_id: profiles.pluck(:id)))
+      .latest
```

Comment:
I'd argue we should add this as a relationship:

```suggestion
    has_many :latest_test_results, -> { latest }, through: :profiles, source: :test_results
```

Reply:
Yeah, this doesn't work as I thought it would work :sweat_smile: